deepview_profile/profiler/utilization.py

```diff
@@ -0,0 +1,492 @@
+                span_end = cuda_slice[0]['ts'] + cuda_slice[0]['dur']
+
+        # check if there is a span_end value otherwise set span to time
+        node.gpu_forward_span = span_end - span_start if span_end != 0 else time
+        node.gpu_forward = time
+
+    def _backward_slices(self,tp):
+        res = []
+        backwardTopOps = tp.query_dict(
+            "select * from slices where name like '%Autograd::engine%' and depth=0 ORDER BY ts ASC")
+        for bto in backwardTopOps:
+            kernel_slices = []
+            endTime = bto['ts'] + bto['dur']
+            cudaLaunchList = tp.query_dict(f"""select * from slices where name like '%CudaLaunchKernel%' and track_id={bto['track_id']}
+                                    and depth>{bto['depth']} and ts between {bto['ts']} and {endTime}""")
+
+            for cudaLaunch in cudaLaunchList:
+                slice_id_origin = cudaLaunch['slice_id']
+                slice_id_destination = tp.query_dict(
+                    f"""select * from flow where slice_out={slice_id_origin}""")
+                if slice_id_destination:
+                    cuda_slice = tp.query_dict(
+                        f"select * from slice where slice_id={slice_id_destination[0]['slice_in']}")
+                    kernel_slices.append(cuda_slice[0])
+
+            bto['name'] = bto["name"].split()[-1]
+            bto['kernel_list'] = kernel_slices
+            res.append(bto)
+        return res
+
+    def _accumulate_backward_slices_to_node(self,node):
+        for ch in node.children:
+            node.cpu_backward_slices.extend(self._accumulate_backward_slices_to_node(ch))
+
+        return node.cpu_backward_slices
+    
+    def _calculate_backward_gpu_span(self,kernel_slices):
+        # return gpu_span, gpu_net_time
+        minTs = float('inf')
+        maxTs = float('-inf')
+        netTime = 0
+        if not kernel_slices:
+            return 0, 0
+        kernel_slices.sort(key=lambda x: x['ts'])
+        for kernel in kernel_slices:
+            minTs = min(minTs, kernel['ts'])
+            maxTs = max(maxTs, kernel['ts']+kernel['dur'])
+            netTime += kernel['dur']
+
+        span = maxTs - minTs
+        return span, netTime
+
+    def _calculate_backward_cpu_span(self,node):
+        backward_slices = node.cpu_backward_slices
+        kernel_slices = []
+        if not backward_slices:
+            return
+        minTs = float('inf')
+        maxTs = float('-inf')
+        netTime = 0
+        backward_slices.sort(key=lambda x: x['ts'])
+        for bw_slice in backward_slices:
+            minTs = min(minTs, bw_slice['ts'])
+            maxTs = max(maxTs, bw_slice['ts']+bw_slice['dur'])
+            netTime += bw_slice['dur']
+            kernel_slices.extend(bw_slice['kernel_list'])
+
+        node.cpu_backward_span = maxTs - minTs
+        node.cpu_backward = netTime
+        node.gpu_backward_span, node.gpu_backward = self._calculate_backward_gpu_span(
```

Comment:
I don't think calculating backward GPU span and backward CPU span should be coupled. I think you can split this function into a few parts. One part can traverse the backward slices and get the kernel slices

You can also use list comprehension to help you out with getting the values of minTs/maxTs/netTime. See https://stackoverflow.com/questions/52468270/comprehension-to-find-the-min-in-a-dict for an example

Reply:
I am not sure, I believe they are decoupled. I have two functions _calculate_backward_cpu_span and _calculate_backward_gpu_span. In the function _backward_slices I add all the gpu kernel calls that belong to a cpu backward call. The reason I am calling _calculate_backward_gpu_span from _calculate_backward_cpu_span  is because I already have all the kernel calls that belong to that cpu slice (kernel_slices)

List comprehension will be a good alternative, but because I am using the loop to append all the kernel calls it is more convenient to process all the information in a single loop.